File: java/09_pgm_avion_interface/src/co/edu/campusucc/poo/implement/TipoMilitar.java
package co.edu.campusucc.poo.implement;

import co.edu.campusucc.poo.abst.Avion;
import co.edu.campusucc.poo.interfaces.IAvionMilitar;

public class TipoMilitar extends Avion implements IAvionMilitar {

    public TipoMilitar(){
        super();
        setTipo("Militar");
    }
    
    @Override
    public void startingEngine() {
        System.out.println("⛽Check Fuel ✅");
        for (int i = 0; i < 3; i++) {
            System.out.println("💥💥Spark💥💥");
            try {
                Thread.sleep(800 - (i * 400));
            } catch (InterruptedException e) {
                System.err.println("⛔:" + e);
            }
        }
        System.out.println("Started the Motor...✈️✈️");
    }

    @Override
    public void speedUp() {
        String speedUpString = "speed up in Progress: ";
        for (int i = 0; i < super.getVelocidad(); i++) {
            speedUpString += "✈️";
        }

        for (int i = 0; i < 10; i++) {
            speedUpString += "✈️";
            System.out.println(speedUpString);
            try {
                Thread.sleep(500 - (i * 50));
            } catch (InterruptedException e) {
                System.err.println("⛔:" + e);
            }
        }
        super.addSpeed(10);
    }

    @Override
    public void stopingEngine() {
        System.out.println("✅ Check the Turbine 🆗🆗");
        for (int i = 0; i < 2; i++) {
            System.out.println("⛔ turning off turbine❌❌");
            try {
                Thread.sleep(800 - (i * 400));
            } catch (InterruptedException e) {
                System.err.println("⛔:" + e);
            }
        }
        System.out.println("Stoped...⚡⚡");
    }

    public void shootMissiles(){
        for (int i = 0; i < 5; i++) {
            System.out.println("🚀🚀 Missiles ready to fire 🔥🔥🔥");
            try {
                Thread.sleep(1000 - (i * 50));
            } catch (InterruptedException e) {
                System.err.println("⛔:" + e);
            }
        }
    }

    @Override
    public void chargeFuel(int litros) {
        super.chargeFuel(litros);
        System.out.println("");
        for (int i = 0; i < litros; i++) {
            try {
                System.out.print("⛽");
                Thread.sleep(100);
            } catch (InterruptedException e) {
                System.err.println("⛔:" + e);
            }
        }
        System.out.println("");
        System.out.println("⛽ Full Fuel ✅");
    }
}

File: c++/09_pgm_avion_interface/src/co/edu/campusucc/poo/implement/TipoMilitar.cpp
#include <iostream>
#include "Avion.h"
#include "IAvionMilitar.h"
#include "TipoMilitar.h"

using namespace std;

TipoMilitar::TipoMilitar() : Avion() {
    setTipo("Militar");
}

void TipoMilitar::startingEngine() {
    cout << "⛽Check Fuel ✅" << endl;
    for (int i = 0; i < 3; i++) {
        cout << "💥💥Spark💥💥" << endl;
        try {
            this_thread::sleep_for(chrono::milliseconds(800 - (i * 400)));
        } catch (const std::exception& e) {
            cerr << "⛔:" << e.what() << endl;
        }
    }
    cout << "Started the Motor...✈️✈️" << endl;
}

void TipoMilitar::speedUp() {
    string speedUpString = "speed up in Progress: ";
    for (int i = 0; i < getVelocidad(); i++) {
        speedUpString += "✈️";
    }

    for (int i = 0; i < 10; i++) {
        speedUpString += "✈️";
        cout << speedUpString << endl;
        try {
            this_thread::sleep_for(chrono::milliseconds(500 - (i * 50)));
        } catch (const std::exception& e) {
            cerr << "⛔:" << e.what() << endl;
        }
    }
    addSpeed(10);
}

void TipoMilitar::stopingEngine() {
    cout << "✅ Check the Turbine 🆗🆗" << endl;
    for (int i = 0; i < 2; i++) {
        cout << "⛔ turning off turbine❌❌" << endl;
        try {
            this_thread::sleep_for(chrono::milliseconds(800 - (i * 400)));
        } catch (const std::exception& e) {
            cerr << "⛔:" << e.what() << endl;
        }
    }
    cout << "Stoped...⚡⚡" << endl;
}

void TipoMilitar::shootMissiles() {
    for (int i = 0; i < 5; i++) {
        cout << "🚀🚀 Missiles ready to fire 🔥🔥🔥" << endl;
        try {
            this_thread::sleep_for(chrono::milliseconds(1000 - (i * 50)));
        } catch (const std::exception& e) {
            cerr << "⛔:" << e.what() << endl;
        }
    }
}

void TipoMilitar::chargeFuel(int litros) {
    Avion::chargeFuel(litros);
    cout << endl;
    for (int i = 0; i < litros; i++) {
        try {
            cout << "⛽";
            this_thread::sleep_for(chrono::milliseconds(100));
        } catch (const std::exception& e) {
            cerr << "⛔:" << e.what() << endl;
        }
    }
    cout << endl;
    cout << "⛽ Full Fuel ✅" << endl;
}
